torch/utils/_pytree.py

```diff
@@ -1181,3 +1300,91 @@ def arg_tree_leaves(*args: PyTree, **kwargs: PyTree) -> List[Any]:
     for a in kwargs.values():
         _tree_leaves_helper(a, leaves)
     return leaves
+
+
+def tree_flatten_with_path(tree: PyTree) -> Tuple[List[Tuple[KeyPath, Any]], TreeSpec]:
+    """Flattens a pytree like :func:`tree_flatten`, but also returns each leaf's key path.
+
+    Args:
+        tree: a pytree to flatten. If it contains a custom type, that type must be
+            registered with an appropriate `tree_flatten_with_path_fn` when registered
+            with :func:`register_pytree_node`.
+    Returns:
+        A tuple where the first element is a list of (key path, leaf) pairs, and the
+        second element is a :class:`TreeSpec` representing the structure of the flattened
+        tree.
+    """
+    _, treespec = tree_flatten(tree)
+    return list(_generate_key_paths((), tree)), treespec
```

Comment:
```suggestion
    treespec = tree_structure(tree)
    return list(_generate_key_paths((), tree)), treespec
```

I'm worried about the consistency between `flatten_fn` (used in `tree_{flatten,map}`) and `flatten_with_keys_fn` (used in `tree_{flatten,map}_with_path`). We need to make sure both functions produce the same `leaves` and `context`.

Another concern is performance, we firstly traverse the whole tree with `flatten_fn` first to get the `treespec`. Then we traverse the whole tree again with `flatten_with_keys_fn` again to get the paths. We can merge the two passes into one.

Reply:
for consistency: I will add the test as mentioned above to check all the types we are registering as part of pytorch for consistency.

For performance: I think this could definitely be improved, but let's leave it for another PR I think. I think it's best to keep PR focused on the basic API design and functional correctness of the implementation, and I think modifying `_generate_key_paths` to also generate a treespec may make that harder.